src/include/rbd/librbd.h

```diff
@@ -1365,6 +1365,30 @@ CEPH_RBD_API int rbd_pool_stats_option_add_uint64(rbd_pool_stats_t stats,
                                                   uint64_t* stat_val);
 CEPH_RBD_API int rbd_pool_stats_get(rados_ioctx_t io, rbd_pool_stats_t stats);
 
+/**
+ * Register a freeze/thaw watcher.
+ *
+ * @param image the image to watch
+ * @param freeze_cb what to do when librbd wants to freeze io
+ * @param thaw_cb what to do when librbd wants to thaw io
+ * @param arg opaque value to pass to the callbacks
+ * @param handle where to store the internal id assigned to this watch
+ * @returns 0 on success, negative error code on failure
+ */
+CEPH_RBD_API int rbd_freeze_thaw_watch(rbd_image_t image,
```

Comment:
Think about this more, I guess we actually need an API that notifies when a snapshot (and/or when a new snapshot discovered during a refresh) and pauses the updating of the snap seq until after the user tells librbd to thaw. The easy solution would be a blocking call to the watcher that when it returns we can assume the image is thawed, but I think it would be cleaner to have an async callback to the client watcher and the client would need to invoke a thaw method to resume. Maybe freeze/thaw is the wrong terminology for this use-case? Maybe it's a `quiesce` callback and a `rbd_quiesce_complete` client->librbd API to resume.

Reply:
Sorry, I seem to fail to understand your idea with discovering a new snapshot during refresh. I think it would mean the snapshot was already created, so why freeze (quiesce) at that point?

Reply:
Just because the snapshot record exists doesn't mean that a remote client can't still write to it. For the case when exclusive-lock is disabled, the snapshot record is created remotely and a header update notification is sent. When the header is refreshed, the remote client sees that a new snapshot has been created and it updates its snap seq to stop writing to the snapshot (which is not R/O at this point).

Reply:
... you could put new freeze/thaw image watcher calls in so that newer clients have a chance to pause the "rbd_snap_create" API call from completing until after all alive clients (that support the new method) have ACKed the freeze.

Reply:
Thanks. It is much clear now.

Reply:
Actually -- thinking about the shared volume use-case, I guess we would want a freeze/quiesce callback (and a freeze/quiesce complete API method) in addition to an unfreeze/unquiesce callback (and associated complete API method) so that we can ensure all clients have completed the freeze/quiesce before allowing any clients to unfreeze/unquiesce. 

This will also be useful for the case when we group multiple images together for consistency so that all images in the group are frozen before they are all unfrozen.